### ISSUE TITLE
[Proposal] Support Automatic "Pretty" Display Name for Test Cases

### ISSUE THREAD

**Issue by commonsensesoftware** [xunit/xunit#759]
## Overview

The **DisplayName** property for facts and theories are great for making tests really easy to read, but after explicitly defining them for a couple of years now, it feels like Xunit could do better at providing an automatic, pretty display using _Convention over Configuration (CoC)_.

For example, many developers write tests as:

``` C#
public class MyTestClass
{
  [Fact]
  public void verify_that_something_does_the_right_thing() {  }
}
```

What the author intended the test to say is "verify that something does the right thing".  This is a limitation in the source language as opposed to an issue with Xunit.  Some Xunit extensions, like that for F#, allow you to write tests in exactly this way.
## Proposal

I propose supporting the automatic pretty display name of classes and methods that follow a well-defined convention. In doing so, display names become easier and more natural to read.

The following rules and conventions would apply:
- Automatic display formatting is enabled by default (IMHO few developers would want to opt out)
- The **DisplayName** property still overrides the default, formatted display name
- The convention rules must be applicable across programming languages
- The underscore ('_') character is always replaced with a space
- The following substitution tokens are supported if they have an underscore ('_') to both sides, appear as the first token, or appear as the last token. Token matches are case-insensitive. These are the same tokens supported in Windows PowerShell:
  - **eq** '='
  - **ne** '!='
  - **lt** '<'
  - **le** '<='
  - **gt** '>'
  - **ge** '>='
- An ASCII escape sequence can be specified using **X[HH]**, where the sequence is uppercase and **H** is a valid hexadecimal character
- A Unicode escape sequence can be specified using **U[HHHH]**, where the sequence is uppercase and **H** is a valid hexadecimal character
## Examples

The following outlines a few basic usage examples:

``` C#
public class Example
{
  [Fact]
  public void unit_tests_are_awesomeX21_U263A(){ }

  [Fact]
  public void api_version_1_should_be_lt_2(){ }

  [Fact]
  public void X3DX3D_operator_overload_should_be_equivalent_to_Equals_method(){ }

  [Fact]
  public void X27stuffedX27_should_not_be_ambiguous_X27stUFFEDX27(){ }

  [Fact]
  public void X27maxed_outX27_should_not_be_ambiguous_with_X27maXED_outX27(){ }

  [Fact]
  public void TestNameShouldRemainUnchanged(){ }

  [Fact( DisplayName = "custom test name" )]
  public void CustomTestName(){ }

  [Fact( DisplayName = "an explicit display name overrides the convention" )]
  public void An_Explicit_DisplayName_Overrides_The_Convention(){ }
}
```

The display name for the facts are:
- unit tests are awesome! ☺
- api version 1 should be < 2
- == operator overload should be equivalent to Equals method
- 'stuffed' should not be ambiguous with 'st￭'
- 'maxed out' should not be ambiguous with 'maí out'
- TestNameShouldRemainUnchanged
- custom test name
- an explicit display name overrides the convention

In _Behavior-Driven Development_ (BDD) with XUnit, the **Given, When, Then (GWT)** approach is also largely popular and this type of naming support would make these tests significantly easier to read by convention. For example:

``` C#
namespace Given
{
  public class a_version_number
  {
    [Fact]
    public void _when_it_equals_1x2C_then_it_should_be_less_than_2() {}
  }
}
```

The display output is **"Given.a version number. when it equals 1, then it should be less than 2"**. Note that the display name has two periods ('.') in it.  These remain due to the current limitation as described in Issue #524. The compromise of a period over a comma in this context is minor and doesn't sacrifice readability or intention. If or when the issue is ever addressed holistically, the naming semantics could be based purely on class and method names. In my experience and option, few developers are concerned with including the namespace in their test case names or descriptions.
## Considerations

I additionally had the following thoughts and considerations:
- Although supporting token substitutions isn't really necessary, it's convenient for very common characters without requiring developers to remember the equivalent hexadecimal. Granted developers still have to remember the monikers, they are more intuitive and my recommendation is to use tokens that are already used in other contexts (e.g. PowerShell).
- Supporting escape sequences isn't a must and is probably not the most common usage scenario; however, I think it's really useful for those that want to take advantage of it. I considered may forms of the escape sequences and the only viable scenario that I could come up with where there would be ambiguity between the atypical text in a class or method name and the escape sequence was allowing a case-insensitive match. If casing isn't considered, some normal text could be confused with an escape sequence; for example: _ma**xed**_ or _st**uffed**_. Both of these words contain matching escape sequences. By requiring escape sequences be all uppercase, the parsing is simpler, faster, and unambiguous, such as _ma**XED**_ or _st**UFFED**_. Given the general guidance around method name casing and the manner in which developers write the names of tests, it seems very unlikely that there would ever be a misinterpretation of intention.
- I considered using only one escape sequence syntax, but in evaluating the implementation it became clear that the disambiguation of **X[HH]**, **X[HHHH]**, and normal text was not clear; hence **U[HHHH]**. I considered only supporting the Unicode syntax, but most usage is likely in the ASCII range, which would force unnecessary extra characters in the name. I also considered only supporting ASCII, but there are a handful of scenarios where someone might want Unicode, in particular, for a symbol glyph.  It turned out that supporting both with unambiguous semantics didn't add any extra complexity to the implementation.
- The performance cost must be negligible and invisible to developers. In my naïve testing, I found the worst cost was an average of 3500 ticks, which is 0.35 milliseconds and likely included JIT. The average time over a set of formatted names was around 15 ticks, which is nearly transparent.
## Conclusion

Given the community contributions, it's unfair to simply ask for this to be implemented.  I've already taken a fork and have an initial cut working with tests. I'd love to get some feedback and submit a pull request if the _Benevolent Dictators_ deem it worthy.

**Comment by jsimmonsdavidccook** [comment]
Is there already support for the logger outputting these test names?

**Comment by commonsensesoftware** [comment]
**Yes**. Any logging that uses **ITestCase.DisplayName** will use the human-readable test display names. There a few exceptions where **ITestCase.UniqueID** or **TestCase.FullyQualifiedName** are used, but that is usually in addition to the **ITestCase.DisplayName**. In terms of logging, console output, and so on, the output is the same as if you were using **FactAttribute.DisplayName**; except that now you don't have to type it twice. =)

**Comment by bradwilson** [comment]
Merged with #828.
